### PR TITLE
AArch64: let ubuntu 20.04+ boot from rust hypervisor firmware

### PR DESCRIPTION
**Pull request by jongwu**
There are several issues block ubuntu 20.04+ boot from rust hypervisor firmware:

1. Grub or UEFI Shim may large enough and overlaps the firmware code region;
2. UEFI Shim or Grub file name length longer than 11 can't be expressed correctly: for example: "BOOTAA64.CSV" will be carried as "BOOTAA64CSV" in rust hypervisor firmware;
3. EFI memory allocations can be ran out;
4. Pass full fdt to OS blocks ACPI initialization;

This PR try to fix them one by one:
1. Move code start address upward to enlarge the memory region used for loading Grub/Shim;
2. Fixup fat file name if needed;
3. Bump allocations max number;
4. Don't pass fdt in system configuration table;

After above fix, ubuntu 20.04+ can boot from rust hypervisor firmware. Test pass for ubuntu 18.04, 20.04 and 22.04.

Fixes: #261

### CODE REVIEW

File: aarch64-unknown-none.ld [hunk 1]
```diff
@@ -6,7 +6,7 @@ ENTRY(ram64_start)
     ACPI:   [0x4020_0000-0x403f_ffff)
     kernel: [0x4048_0000-]
    The stack start is at the end of the DRAM region. */
-ram_min = 0x40480000;
+ram_min = 0x40600000;
 
 SECTIONS
 {
```
Comment (rbradford) [inline aarch64-unknown-none.ld:12]: I don't understand how this gives more room for GRUB. This is file determines the layout of the ELF executable that is compiled from the project. Your commit message seems to imply that the EFI payload is loaded into the memory just below the RHF executable code. I think this is a bit different to how we do it elsewhere. If you want to keep it the same it would be better to.

Make `ram_min = 0x4000000` the true start of RAM and then add a `payload_start/payload_end` region to the linker file to save the memory and push back the code region.

@retrage Perhaps you can clarify on this?

File: src/arch/aarch64/ram64.s [hunk 2]
```diff
@@ -14,8 +14,8 @@ ram64_start:
   add x13, x18, #0x16   /* code0: UEFI "MZ" signature magic instruction */
   b jump_to_rust        /* code1 */
 
-  .quad 0               /* text_offset */
-  .quad 0               /* image_size */
+  .quad 0x200000        /* text_offset */
```
Comment (rbradford) [inline src/arch/aarch64/ram64.s:17]: Could we resolve the symbols from the linker file instead of hardcoding?

**Review by rbradford — CHANGES_REQUESTED** [review]

File: aarch64-unknown-none.ld (hunk 1 above)
Comment (jongwu) [reply aarch64-unknown-none.ld:12]: Hi @rbradford -, In current design, 0x4040000 ~ 0x4048000 is reserved for RHF payload. 0x40000000 ~0x40400000 is reserved for DTB and ACPI.

File: src/main.rs [hunk 3]
```diff
@@ -132,6 +135,12 @@ fn boot_from_device(device: &mut block::VirtioBlockDevice, info: &dyn bootinfo::
         }
     };
 
+    #[cfg(target_arch = "aarch64")]
+    if code_range().start < (info.kernel_load_addr() + size) as usize {
+        log!("error: Boot Image is too large!");
```
[the remark this answers is not in the document]
Comment (jongwu) [reply src/main.rs:140]: What about "log!("Error Boot Image is too large");"

File: aarch64-unknown-none.ld (hunk 1 above)
Comment (rbradford) [reply aarch64-unknown-none.ld:12]: So what does that have to do with the EFI payload?

Comment (rbradford) [reply aarch64-unknown-none.ld:12]: I think it would be better to load the kernel/grub so that it doesn't overwrite the memory that RHF is loaded at e.g. on x86 RHF is loaded at 1MiB and the EFI binary/kernel is loaded at 2MiB. The correct place to change is probably the value provided to `StartInfo::new` for `kernel_load_addr` as that is the address that is passed to the EFI PE loader.

Comment (jongwu) [reply aarch64-unknown-none.ld:12]: I think this is a big change that impact CH, RHF and maybe also future change for EDK2 (I'd like to unify the boot process of EDK2 and RHF), so let me think it through carefully.

File: aarch64-unknown-none.ld [hunk 4]
```diff
@@ -6,7 +6,7 @@ ENTRY(ram64_start)
     ACPI:   [0x4020_0000-0x403f_ffff)
     kernel: [0x4048_0000-]
    The stack start is at the end of the DRAM region. */
```
Comment (retrage) [inline aarch64-unknown-none.ld:8]:
```suggestion
    payload:[0x4040_0000-0x405f_ffff)
    RHF:    [0x40600000-]
   Assuming 2MB is enough to load payload.
   The stack start is at the end of the DRAM region. */
```
Add comments on the memory layout assumption.

**Review by retrage — COMMENTED** [review]
@jongwu Thanks a lot for fixing it to boot Ubuntu. This is a huge milestone in #198.

File: aarch64-unknown-none.ld (hunk 1 above)
Comment (retrage) [reply aarch64-unknown-none.ld:12]: The points are:
* Assume that 2MB of space is enough to load GRUB. To do this:
  * Move `ram_min` to `0x40600000`.
  * Add a run-time payload size check.
I think the change in 6735c61c8929b1697d20d76310635557b0748c44 is little bit ad hoc, but it's acceptable because it works. It's future work to implement better a loading process.

So, my suggestion is to just add comments to the memory layout assumption. See https://github.com/cloud-hypervisor/rust-hypervisor-firmware/pull/262#discussion_r1257456696.

File: src/fat.rs [hunk 5]
```diff
@@ -1172,16 +1185,16 @@ mod tests {
...
         super::name_to_str(".", &mut s);
         assert_eq!(crate::common::ascii_strip(&s), ".");
-        let mut s = [0_u8; 11];
+        let mut s = [0_u8; 12];
         super::name_to_str("..", &mut s);
         assert_eq!(crate::common::ascii_strip(&s), "..");
-        let mut s = [0_u8; 11];
+        let mut s = [0_u8; 12];
         super::name_to_str("ABCDEFGHIJK", &mut s);
         assert_eq!(crate::common::ascii_strip(&s), "ABCDEFGHIJK");
```
Comment (retrage) [inline src/fat.rs:1199]:
```suggestion
        assert_eq!(crate::common::ascii_strip(&s), "ABCDEFGH.IJK");
```
Should we need to fix the test case?

File: aarch64-unknown-none.ld (hunk 1 above)
Comment (jongwu) [reply aarch64-unknown-none.ld:12]: OK

File: src/efi/mod.rs [hunk 6]
```diff
@@ -1115,6 +1114,8 @@ pub fn efi_exec(
     let mut ct_index = 0;
 
     // Populate with FDT table if present
+    // To boot with ACPI, we can't offer full fdt. Simply, neglect it. See https://github.com/torvalds/linux/blob/d528014517f2b0531862c02865b9d4c908019dc4/arch/arm64/kernel/acpi.c#L203
+    #[cfg(not(target_arch = "aarch64"))]
```
[the remark this answers is not in the document]
Comment (jongwu) [reply src/efi/mod.rs:1118]: I think so. Even though for Arm, we should pass a FDT without platform info. See 7.4.3 in [SBBR](https://developer.arm.com/documentation/den0044/latest). What I do here is just a workaround. I will open another PR to improve it.

File: aarch64-unknown-none.ld (hunk 4 above)
Comment (jongwu) [reply aarch64-unknown-none.ld:8]: OK.

**Review by retrage — APPROVED** [review]
It looks good to me. Thanks!

File: src/fat.rs [hunk 7]
```diff
@@ -232,6 +232,17 @@ fn name_to_str(input: &str, output: &mut [u8]) {
             break;
         }
     }
+
+    // Fixup for Director or file name, when the length of its name and extend name equals 11. There maybe no
```
Comment (rbradford) [inline src/fat.rs:236]: nit: typo here - I think you mean directory. Could you also wrap you comment to 80 characters

File: src/efi/mod.rs [hunk 8]
```diff
@@ -1115,6 +1114,8 @@ pub fn efi_exec(
     let mut ct_index = 0;
 
     // Populate with FDT table if present
+    // To boot with ACPI, we can't offer full fdt. Simply, neglect it. See https://github.com/torvalds/linux/blob/d528014517f2b0531862c02865b9d4c908019dc4/arch/arm64/kernel/acpi.c#L203
```
Comment (rbradford) [inline src/efi/mod.rs:1117]: Could you wrap this comment at 80 characters (or close - by putting the URL on a new line of comment.

File: src/efi/mod.rs [hunk 9]
```diff
@@ -614,7 +614,6 @@ pub extern "efiapi" fn install_configuration_table(guid: *mut Guid, table: *mut
 
     for entry in ct.iter_mut() {
         if entry.vendor_guid == unsafe { *guid } {
-            entry.vendor_table = table;
```
Comment (rbradford) [inline src/efi/mod.rs:617]: This should be a separate commit.

File: src/efi/mod.rs [hunk 10]
```diff
@@ -1115,6 +1114,8 @@ pub fn efi_exec(
     let mut ct_index = 0;
 
     // Populate with FDT table if present
```
Comment (rbradford) [inline src/efi/mod.rs:1116]: Could you update the commit title to mark as aarch64 specific.

**Review by rbradford — COMMENTED** [review]
Thanks @retrage for the review - just a few bit from me!

**Comment by jongwu** [comment]
Thanks @rbradford -, all comments addressed

**Review by rbradford — COMMENTED** [review]
I'll let you have a go at improving the other commit messages in the series. Let me know if you want help.

File: aarch64-unknown-none.ld [hunk 11]
```diff
@@ -4,9 +4,17 @@ ENTRY(ram64_start)
...
     FDT:    [0x4000_0000-0x401f_ffff)
     ACPI:   [0x4020_0000-0x403f_ffff)
-    kernel: [0x4048_0000-]
-   The stack start is at the end of the DRAM region. */
-ram_min = 0x40480000;
+    payload:[0x4040_0000-0x405f_ffff)
+    RHF:    [0x40600000-]
+   Assuming 2MB is enough to load payload.
+   The stack start is at the end of the RHF region. */
+ram_min = 0x40600000;
```
[the remark this answers is not in the document]
Comment (jongwu) [reply aarch64-unknown-none.ld:11]: Thanks @rbradford -, done.

**Review by rbradford — APPROVED** [review]
I worked on your commit messages @jongwu - please let me know if you're happy with those changes!

**Comment by jongwu** [comment]
@rbradford -, Great! Thanks! 😄
